client/jmx-adapter/src/main/java/org/jolokia/client/jmxadapter/RemoteJmxAdapter.java

```diff
@@ -0,0 +1,586 @@
+        result.add(objectName);
+      }
+    }
+
+    return result;
+  }
+
+  private boolean applyQuery(QueryExp query, ObjectName objectName) {
+    if (QueryEval.getMBeanServer() == null) {
+      query.setMBeanServer(this.createStandInMbeanServerProxyForQueryEvaluation());
+    }
+    try {
+      return query.apply(objectName);
+    } catch (RuntimeException e) {
+      throw e;
+    } catch (Exception e) {
+      throw new UncheckedJmxAdapterException(e);
+    }
+  }
+
+  /**
+   * Query evaluation requires an MBeanServer but only to figure out objectInstance
+   *
+   * @return a dynamic proxy dispatching getObjectInstance back to myself
+   */
+  private MBeanServer createStandInMbeanServerProxyForQueryEvaluation() {
+    return (MBeanServer)
+        Proxy.newProxyInstance(
+            this.getClass().getClassLoader(),
+            new Class[]{MBeanServer.class},
+            new InvocationHandler() {
+              @Override
+              public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
+                if (method.getName().contains("getObjectInstance") && args.length == 1) {
+                  return getObjectInstance((ObjectName) args[0]);
+                } else {
+                  throw new UnsupportedOperationException(
+                      "This MBeanServer proxy does not support " + method);
+                }
+              }
+            });
+  }
+
+  @SuppressWarnings("unchecked")
+  private <RESP extends J4pResponse<REQ>, REQ extends J4pRequest> RESP unwrapExecute(REQ pRequest)
+      throws IOException, InstanceNotFoundException {
+    try {
+      pRequest.setPreferredHttpMethod("POST");
+      return this.connector.execute(pRequest, defaultProcessingOptions());
+    } catch (J4pException e) {
+      return (RESP) unwrapException(e);
+    }
+  }
+
+  private Map<J4pQueryParameter, String> defaultProcessingOptions() {
+    if (this.defaultProcessingOptions == null) {
+      this.defaultProcessingOptions = new HashMap<J4pQueryParameter, String>();
+      defaultProcessingOptions.put(J4pQueryParameter.MAX_DEPTH, "10"); // in case of stack overflow
+      defaultProcessingOptions.put(J4pQueryParameter.SERIALIZE_EXCEPTION, "true");
+    }
+    return defaultProcessingOptions;
+  }
+
+  private static final Set<String> UNCHECKED_REMOTE_EXCEPTIONS =
+      Collections.singleton("java.lang.UnsupportedOperationException");
+
+  @SuppressWarnings("rawtypes")
+  private J4pResponse unwrapException(J4pException e)
+      throws IOException, InstanceNotFoundException {
+    if (e.getCause() instanceof IOException) {
+      throw (IOException) e.getCause();
+    } else if (e.getCause() instanceof RuntimeException) {
+      throw (RuntimeException) e.getCause();
+    } else if (e.getCause() instanceof Error) {
+      throw (Error) e.getCause();
+    } else if (e.getMessage()
+        .matches("Error: java.lang.IllegalArgumentException : No MBean '.+' found")) {
```

Comment:
Maybe matching only on "No MBean .+ found" is good enough, to be less fragile on upstream changes in the error message (as this is a very subtle dependency that no one is aware of)

Reply:
I wonder whether we could do better in the backend to explicitly declare that an MBean is not found ?